File: go/array/leetcode/three_sum.go
package main

import (
	"sort"
)

//三数之和

//暴力求解
func threeSum1(nums []int) [][]int {
	res := make([][]int, 0)
	set := map[[3]int]bool{}

	for i := 0; i < len(nums)-2; i++ {
		for j := i + 1; j < len(nums)-1; j++ {
			for k := j + 1; k < len(nums); k++ {
				if nums[i]+nums[j] == -nums[k] {
					lis := []int{nums[i], nums[j], nums[k]}
					sort.Ints(lis)

					var arr [3]int
					copy(arr[:], lis)
					if !set[arr] {
						res = append(res, lis)
						set[arr] = true
					}
				}
			}
		}
	}
	return res
}

//排序后，双指针解法
func threeSum2(nums []int) [][]int {
	var res [][]int
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if nums[i] > 0 {
			break
		}
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		j, k := i+1, len(nums)-1
		for j < k {
			if nums[j]+nums[k] == -nums[i] {
				res = append(res, []int{nums[i], nums[j], nums[k]})
				j++
				k--
				for j < k && nums[j] == nums[j-1] {
					j++
				}
				for j < k && nums[k] == nums[k+1] {
					k--
				}
			} else if nums[j]+nums[k] < -nums[i] {
				j++
			} else {
				k--
			}
		}
	}
	return res
}

File: python/array/leetcode/three_sum.py
from typing import List

# 三数之和


class Solution:

    # 暴力法，三层循环
    # 超出时间限制
    def threeSum_1(self, nums: List[int]) -> List[List[int]]:
        res, s = [], set()
        for i in range(len(nums) - 2):
            for j in range(i + 1, len(nums) - 1):
                for k in range(j + 1, len(nums)):
                    if nums[i] + nums[j] == -nums[k]:
                        t = tuple(sorted((nums[i], nums[j], nums[k])))
                        if t not in s:
                            res.append([nums[i], nums[j], nums[k]])
                            s.add(t)
        return res

    # 排序之后，使用双指针
    def threeSum_2(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i, v in enumerate(nums):
            if v > 0:
                break
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            j, k = i + 1, len(nums) - 1
            while j < k:
                if nums[j] + nums[k] < -v:
                    j += 1
                elif nums[j] + nums[k] > -v:
                    k -= 1
                else:
                    res.append([nums[i], nums[j], nums[k]])
                    j, k = j + 1, k - 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
        return res


if __name__ == '__main__':
    print('a')